akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2285,6 +2285,281 @@ private[akka] final class StatefulMap[S, In, Out](
+      }
+
+      override def postStop(): Unit = {
+        state match {
+          case Some(value) =>
+            // outlet isn't available here
+            if (stateAcquired) onComplete(value)
+          //TODO figure out what should be done here
+          // since resource maybe in the half way of acquisition
+          case None =>
+        }
+      }
+
+      override def preStart(): Unit = {
+        initState()
+        buffer = BufferImpl(parallelism, inheritedAttributes)
+      }
+
+      override def onPull(): Unit = pushNextIfPossible()
+
+      override def onPush(): Unit = {
+        val elem = grab(in)
+        state match {
+          case Some(_) =>
+            applyUserFunction(elem)
+            pullIfNeeded()
+          case None =>
+            earlyPulledElem = Some(elem)
+            pushNextIfPossible()
+        }
+      }
+
+      @tailrec
+      private def pushNextIfPossible(): Unit =
+        if (state.isEmpty) tryPushAfterInitialized = true
+        else if (earlyPulledElem.isDefined) applyUserFunction(earlyPulledElem.get)
+        else if (buffer.isEmpty) pullIfNeeded()
+        else if (buffer.peek().elem eq NotYetThere) pullIfNeeded() // ahead of line blocking to keep order
+        else if (isAvailable(out)) {
+          val holder = buffer.dequeue()
+          holder.elem match {
+            case Success(elem) =>
+              if (elem != null && elem._1 != null && elem._2 != null) {
+                // state has to be presented here
+                state = Some(combineState(state.get, elem._1))
+                push(out, elem._2)
+                pullIfNeeded()
+              } else {
+                // elem is null
+                pullIfNeeded()
+                pushNextIfPossible()
+              }
+
+            case Failure(NonFatal(ex)) =>
+              holder.supervisionDirectiveFor(decider, ex) match {
+                // this could happen if we are looping in pushNextIfPossible and end up on a failed future before the
+                // onComplete callback has run
+                case Supervision.Stop => failStage(ex)
+                case _                =>
+                  // try next element
+                  pushNextIfPossible()
+              }
+            case Failure(ex) =>
+              // fatal exception in buffer, not sure that it can actually happen, but for good measure
+              throw ex
+          }
+        }
+
+      private def pullIfNeeded(): Unit = {
+        // this will only be called after state future was completed
+        // so no need to check if state future is still pending or first element buffer is still full
+        if (isClosed(in) && buffer.isEmpty) releaseThenCompleteStage()
+        else if (earlyPulledElem.isEmpty && buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
+        // else already pulled and waiting for next element
+      }
+
+      def initState(): Unit = {
+        val future = create()
```

Comment:
this `create` can fail, I think the behavior of `statefulMap` need adjusting too.